app/src/androidTest/java/ch/epfl/balelecbud/authentication/MockAuthenticator.java

```diff
@@ -32,169 +25,75 @@
         }
     };
 
+    private static int uid = 0;
+
+    private User currentUser;
+
     private MockAuthenticator() {
         loggedIn = false;
     }
 
     @Override
-    public void signIn(final String email, final String password, Callback callback) {
+    public CompletableFuture<User> signIn(final String email, final String password) {
         if (users.containsKey(email) && users.get(email).equals(password)) {
             setLoggedIn(true);
-            callback.onSuccess();
+            return MockDatabaseWrapper.getInstance().getDocument("users", "0", User.class).thenCompose(new Function<User, CompletionStage<User>>() {
+                @Override
+                public CompletionStage<User> apply(User user) {
+                    currentUser = user;
+                    return CompletableFuture.completedFuture(user);
+                }
+            });
         } else {
-            callback.onFailure("Login failed");
+            return CompletableFuture.completedFuture(null).thenApply(new Function<Object, User>() {
+                @Override
+                public User apply(Object o) {
+                    throw new RuntimeException("Failed login");
+                }
+            });
         }
     }
 
     @Override
-    public void createAccount(final String email, final String password, Callback callback) {
+    public CompletableFuture<Void> createAccount(final String email, final String password) {
         if (!users.containsKey(email)) {
             users.put(email, password);
             setLoggedIn(true);
-            callback.onSuccess();
+            User u = new User(email, email, String.valueOf(uid));
+            uid++;
+            MockDatabaseWrapper.getInstance().storeDocumentWithID("users", String.valueOf(uid), u);
+            return CompletableFuture.completedFuture(null);
         } else {
-            callback.onFailure("Registration failed: account already exists with this email");
+            return CompletableFuture.completedFuture(null).thenApply(new Function<Object, Void>() {
```

Comment:
Weird that there is no way to create an ExceptionnalFuture but I also looked in the interface and did not find a cleaner way. Might be nice to hide in an util function though ?

Reply:
You're right. A `CompletableFutureUtil` class would be nice to have in order to provide such futures. Thanks !